### PR TITLE
[mvt] implement step expressions + improve conversion of sprites

### LINKED ISSUE #57989: Mapbox GL style converter skips `step` expressions
### What is the bug or the crash?

When opening an MVT layer with a Mapbox or MapLibre stylesheet, QGIS attempts to convert the stylesheet to QGIS symbology, but this converter lacks support for the very common [`step`](https://maplibre.org/maplibre-style-spec/expressions/#step) expression operator, causing significant glitches in the resulting layer.

### Steps to reproduce the issue

1. Follow [these instructions](https://wiki.openstreetmap.org/wiki/OpenHistoricalMap/QGIS#Vector_tiles) to create a vector tile layer that has many symbols and labels based on an OpenHistoricalMap style.
2. Note the banner warning that “Style could not be completely converted”. Click the Details button.

Many of the layers include properties or filters set to `step` expressions, which result in the following warnings:

> * roads_motorway: Skipping unsupported expression
> * roads_motorway_construction: Skipping unsupported expression
> * roads_motorwaybridgetop: Skipping unsupported expression
> * roads_motorwaybridgetop_construction: Skipping unsupported expression

### Versions

(Yes, I know v3.36 is a couple releases behind, but it’s the most recent version currently available for macOS.)

QGIS version | 3.36.3-Maidenhead | QGIS code revision | 2df9655469b
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.5
GDAL/OGR version | 3.3.2
PROJ version | 8.1.1
EPSG Registry database version | v10.028 (2021-07-07)
GEOS version | 3.9.1-CAPI-1.14.2
SQLite version | 3.35.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.11.5
OS version | macOS 14.5
  |   |   |  
Active Python plugins
QuickOSM | 1.16.0
nominatim_locator_filter | 0.2.3
processing | 2.12.99
grassprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.6

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [ ] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

The warnings come from here:

https://github.com/qgis/QGIS/blob/5e53e183a5da95b68c4d8a259109cb02ce2cdaba/src/core/vectortile/qgsmapboxglstyleconverter.cpp#L3110

The converter has some support for the `interpolate` operator, so `step` should be handled similarly.

### PR DESCRIPTION
**Pull request by 3nids**
this implements step expressions

it also fixes some cases where the sprite was not defined (mainly when data defined) + support step definition

see https://docs.mapbox.com/style-spec/reference/expressions/#step

fixes #57989

### CODE REVIEW

File: src/core/vectortile/qgsmapboxglstyleconverter.h [hunk 1]
```diff
@@ -756,7 +756,21 @@ class CORE_EXPORT QgsMapBoxGlStyleConverter
...
+      QString spriteSizeProperty;
+      return retrieveSpriteAsBase64WithProperties( value, context, spriteSize, spriteProperty, spriteSizeProperty );
+    }
+
+    /**
+     * Retrieves the sprite image with the specified \a name, taken from the specified \a context as a base64 encoded value
+     *
+     * The \a context must have valid sprite definitions and images set via QgsMapBoxGlStyleConversionContext::setSprites()
+     * prior to conversion.
+     */
```
Comment (nirvn) [inline src/core/vectortile/qgsmapboxglstyleconverter.h:772]: \since

File: tests/src/python/test_qgsmapboxglconverter.py [hunk 2]
```diff
@@ -217,9 +219,25 @@ def testParseMatchList(self):
             -1,
             "rgb(248,249,250)",
             "rgb(252, 252, 252)"
-        ], QgsMapBoxGlStyleConverter.PropertyType.Color, conversion_context, 2.5, 200)
```
Comment (nyalldawson) [inline tests/src/python/test_qgsmapboxglconverter.py:220]: why was this test dropped?

File: tests/src/python/test_qgsmapboxglconverter.py [hunk 3]
```diff
@@ -1428,6 +1446,59 @@ def testFillColorDDHasBrush(self):
...
 
+    def testRetrieveSrpite(self):
```
Comment (nyalldawson) [inline tests/src/python/test_qgsmapboxglconverter.py:1449]:
```suggestion
    def testRetrieveSprite(self):
```

File: tests/src/python/test_qgsmapboxglconverter.py (hunk 2 above)
Comment (3nids) [reply tests/src/python/test_qgsmapboxglconverter.py:220]: this is an error, I changed it back

**Comment by 3nids** [comment]
This is ready to merge.

File: src/core/vectortile/qgsmapboxglstyleconverter.cpp [hunk 4]
```diff
@@ -2743,8 +2747,14 @@ QgsProperty QgsMapBoxGlStyleConverter::parseMatchList( const QVariantList &json,
 
     }
 
-    caseString += QStringLiteral( "WHEN %1 IN (%2) THEN %3 " ).arg( attribute,
-                  matchString.join( ',' ), valueString );
+    if ( matchString.count() == 1 )
+    {
+      caseString += QStringLiteral( "WHEN %1 IS %2 THEN %3 " ).arg( attribute, matchString.at( 0 ), valueString );
```
Comment (nirvn) [inline src/core/vectortile/qgsmapboxglstyleconverter.cpp:2752]: Thanks for the optimization.

File: src/core/vectortile/qgsmapboxglstyleconverter.h [hunk 5]
```diff
@@ -669,6 +669,14 @@ class CORE_EXPORT QgsMapBoxGlStyleConverter
     static QgsProperty parseMatchList( const QVariantList &json, PropertyType type, QgsMapBoxGlStyleConversionContext &context, double multiplier = 1,
                                        int maxOpacity = 255, QColor *defaultColor SIP_OUT = nullptr, double *defaultNumber SIP_OUT = nullptr );
 
+    /**
+     * Parses and converts a match function value list.
+     *
+     * \warning This is private API only, and may change in future QGIS versions
```
Comment (nirvn) [inline src/core/vectortile/qgsmapboxglstyleconverter.h:675]: Do we skip \since when it's a private API even though it's exposed to python?

**Review by nirvn — APPROVED** [review]
@3nids , glad this is being addressed -- only one doc comment.

**Pull request merged**